forgevtt-module.js

```diff
@@ -1872,80 +1879,96 @@ class ForgeVTT_FilePicker extends FilePicker {
+                    return ["forgevtt", forgePath, bucket.key];
                 }
             }
             // Fallback - we weren't able to find the correct bucket. Default to our own assets library (or
             // the custom key, for local installs).
             // Technically this a side effect, but we need to set the bucket label here
             // since the caller doesn't.
-            this.sources.forgevtt.bucket = this._forgeBucketIndex[0].key;
+            this.sources.forgevtt.bucket = userBucket.key;
+            return ["forgevtt", "", undefined];
+        }
+        if (!target) {
             return ["forgevtt", "", undefined];
         }
-        if (!target)
-            return ["forgevtt", ""];
-        // Note: we don't need to insert the `undefined` third return value here.
-        // It is inserted in earlier returns for clarity only.
-
         if (ForgeVTT.usingTheForge) {
             // If not an assets URL but the path is not a known core data folder and isn't a module or system folder
             // then we can assume that it won't be a folder that exists in data and we can infer the source as being
             // from the assets library, even if it's a relative path
             const dataDirs = ["systems", "modules"];
             const publicDirs = ["cards", "icons", "sounds", "ui"];
             if ([...dataDirs, ...publicDirs].every((folder) => !target.startsWith(`${folder}/`))) {
-                return ["forgevtt", target];
+                return ["forgevtt", target, undefined];
             }
         }
-        return super._inferCurrentDirectory(target);
+        return super._inferCurrentDirectory(_target);
     }
-    /* eslint-enable no-param-reassign */
 
+    /**
+     * Determines whether the user can upload assets.
+     * @returns {boolean} Returns true if the user can upload assets, otherwise false.
+     */
     get canUpload() {
         if (this.activeSource === "forgevtt") {
-            if (this.source.bucket === "my-assets") {
-                return true;
-            }
-            const bucket = this._forgeBucketIndex.find((b) => b.key === this.source.bucket);
-            const apiKey = bucket ? bucket.jwt : null;
-            if (!ForgeAPI.isValidAPIKey(apiKey)) {
-                return false;
+            const bucket = this.constructor._getForgeVttBucket(this.source.bucket);
+            if (bucket) {
+                if (bucket.key === "my-assets") {
+                    return true;
+                }
+                if (!ForgeAPI.isValidAPIKey(bucket.jwt)) {
+                    return false;
+                }
+                const permissions = ForgeAPI._tokenToInfo(bucket.jwt).permissions || [];
+                return permissions.includes("write-assets");
             }
-            const info = ForgeAPI._tokenToInfo(apiKey);
-            return (info.permissions || []).includes("write-assets");
         }
```

Comment:
I think this section needs a `return false` after the close of the `if (bucket)` block, at least in order to match up with the old logic paths. Previously, if the bucket was not found then it would return false because `ForgeAPI.isValidAPIKey(null)` would return false. In your version, that check is skipped if the `bucket` is not found, which will let this fall back to default Foundry behaviour.